File: libraries/@mattermost/rnutils/android/src/main/java/com/mattermost/rnutils/session_attributes/SessionAttributesStore.kt
package com.mattermost.rnutils.session_attributes

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest

data class SAField(
    val name: String,
    val type: String,
    val ttlSeconds: Int,
    val gracePeriodSeconds: Int,
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("name", name)
            put("type", type)
            put("ttl_seconds", ttlSeconds)
            put("grace_period_seconds", gracePeriodSeconds)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): SAField? {
            val name = json.optString("name", "")
            val type = json.optString("type", "")
            if (name.isEmpty() || type.isEmpty()) {
                return null
            }
            return SAField(
                name = name,
                type = type,
                ttlSeconds = json.optInt("ttl_seconds", 0),
                gracePeriodSeconds = json.optInt("grace_period_seconds", 0),
            )
        }
    }
}

data class ServerSessionAttributesState(
    var enabled: Boolean,
    var manifest: MutableList<SAField>,
    var lastSentAt: MutableMap<String, Long>,
) {
    fun toJson(): JSONObject {
        val manifestArray = JSONArray()
        manifest.forEach { manifestArray.put(it.toJson()) }
        val lastSent = JSONObject()
        lastSentAt.forEach { (key, value) -> lastSent.put(key, value) }
        return JSONObject().apply {
            put("enabled", enabled)
            put("manifest", manifestArray)
            put("lastSentAt", lastSent)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): ServerSessionAttributesState {
            val manifest = mutableListOf<SAField>()
            val manifestArray = json.optJSONArray("manifest")
            if (manifestArray != null) {
                for (i in 0 until manifestArray.length()) {
                    SAField.fromJson(manifestArray.getJSONObject(i))?.let { manifest.add(it) }
                }
            }
            val lastSentAt = mutableMapOf<String, Long>()
            val lastSent = json.optJSONObject("lastSentAt")
            if (lastSent != null) {
                lastSent.keys().forEach { key ->
                    lastSentAt[key] = lastSent.optLong(key)
                }
            }
            return ServerSessionAttributesState(
                enabled = json.optBoolean("enabled", false),
                manifest = manifest,
                lastSentAt = lastSentAt,
            )
        }
    }
}

class SessionAttributesStore(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences(
        SessionAttributesConstants.PREFS_NAME,
        Context.MODE_PRIVATE,
    )

    fun serverKey(serverUrl: String): String {
        val normalized = serverUrl.trimEnd('/')
        val digest = MessageDigest.getInstance("SHA-256").digest(normalized.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun stateKey(serverUrl: String): String {
        return "${SessionAttributesConstants.STORE_PREFIX}state_${serverKey(serverUrl)}"
    }

    fun loadState(serverUrl: String): ServerSessionAttributesState? {
        val raw = prefs.getString(stateKey(serverUrl), null) ?: return null
        return try {
            ServerSessionAttributesState.fromJson(JSONObject(raw))
        } catch (_: Exception) {
            null
        }
    }

    fun saveState(serverUrl: String, state: ServerSessionAttributesState) {
        prefs.edit().putString(stateKey(serverUrl), state.toJson().toString()).apply()
    }

    fun removeState(serverUrl: String) {
        prefs.edit().remove(stateKey(serverUrl)).apply()
    }

    fun setEnabled(serverUrl: String, enabled: Boolean) {
        val state = loadState(serverUrl) ?: ServerSessionAttributesState(false, mutableListOf(), mutableMapOf())
        state.enabled = enabled
        if (!enabled) {
            state.manifest.clear()
            state.lastSentAt.clear()
        }
        saveState(serverUrl, state)
    }

    fun setManifest(serverUrl: String, manifest: List<SAField>) {
        val state = loadState(serverUrl) ?: ServerSessionAttributesState(true, mutableListOf(), mutableMapOf())
        state.enabled = true
        state.manifest = manifest.toMutableList()
        state.lastSentAt.clear()
        saveState(serverUrl, state)
    }

    fun upsertField(serverUrl: String, field: SAField) {
        val state = loadState(serverUrl)?.takeIf { it.enabled } ?: return
        val index = state.manifest.indexOfFirst { it.name == field.name }
        if (index == -1) {
            state.manifest.add(field)
        } else {
            state.manifest[index] = field
        }
        state.lastSentAt.remove(field.name)
        saveState(serverUrl, state)
    }

    fun removeField(serverUrl: String, name: String) {
        val state = loadState(serverUrl)?.takeIf { it.enabled } ?: return
        state.manifest.removeAll { it.name == name }
        state.lastSentAt.remove(name)
        saveState(serverUrl, state)
    }

    fun setStableValues(values: Map<String, String>) {
        val json = JSONObject()
        values.forEach { (key, value) -> json.put(key, value) }
        prefs.edit().putString(SessionAttributesConstants.STABLE_VALUES_KEY, json.toString()).apply()
    }

    fun getStableValue(name: String): String? {
        val raw = prefs.getString(SessionAttributesConstants.STABLE_VALUES_KEY, null) ?: return null
        return try {
            val value = JSONObject(raw).optString(name, "")
            if (value.isEmpty()) null else value
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/Gekidou/Sources/Gekidou/SessionAttributes/SessionAttributesStore.swift
//
//  SessionAttributesStore.swift
//  Gekidou
//
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
//

import Foundation
import CryptoKit

struct SAField: Codable, Equatable {
    let name: String
    let type: String
    let ttl_seconds: Int
    let grace_period_seconds: Int
}

struct ServerSessionAttributesState: Codable {
    var enabled: Bool
    var manifest: [SAField]
    var lastSentAt: [String: Double]
}

public class SessionAttributesStore {
    public static let shared = SessionAttributesStore()

    private let preferences = Preferences.default
    private let queue = DispatchQueue(label: "com.mattermost.sessionattributes.store")

    private init() {}

    func serverKey(_ serverUrl: String) -> String {
        let normalized = serverUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let digest = SHA256.hash(data: Data(normalized.utf8))
        return digest.compactMap { String(format: "%02x", $0) }.joined()
    }

    private func stateKey(for serverUrl: String) -> String {
        return "\(SessionAttributesConstants.storePrefix)state_\(serverKey(serverUrl))"
    }

    private func readState(for serverUrl: String) -> ServerSessionAttributesState? {
        guard let data = preferences.object(forKey: stateKey(for: serverUrl)) as? Data else {
            return nil
        }
        return try? JSONDecoder().decode(ServerSessionAttributesState.self, from: data)
    }

    private func writeState(_ state: ServerSessionAttributesState, for serverUrl: String) {
        guard let data = try? JSONEncoder().encode(state) else {
            return
        }
        preferences.set(data, forKey: stateKey(for: serverUrl))
    }

    func loadState(for serverUrl: String) -> ServerSessionAttributesState? {
        queue.sync {
            readState(for: serverUrl)
        }
    }

    func saveState(_ state: ServerSessionAttributesState, for serverUrl: String) {
        queue.sync {
            writeState(state, for: serverUrl)
        }
    }

    func removeState(for serverUrl: String) {
        queue.sync {
            preferences.removeObject(forKey: stateKey(for: serverUrl))
        }
    }

    func setEnabled(_ enabled: Bool, for serverUrl: String) {
        queue.sync {
            var state = readState(for: serverUrl) ?? ServerSessionAttributesState(enabled: false, manifest: [], lastSentAt: [:])
            state.enabled = enabled
            if !enabled {
                state.manifest = []
                state.lastSentAt = [:]
            }
            writeState(state, for: serverUrl)
        }
    }

    func setManifest(_ manifest: [SAField], for serverUrl: String) {
        queue.sync {
            var state = readState(for: serverUrl) ?? ServerSessionAttributesState(enabled: true, manifest: [], lastSentAt: [:])
            state.enabled = true
            state.manifest = manifest
            state.lastSentAt = [:]
            writeState(state, for: serverUrl)
        }
    }

    func upsertField(_ field: SAField, for serverUrl: String) {
        queue.sync {
            guard var state = readState(for: serverUrl), state.enabled else {
                return
            }
            if let index = state.manifest.firstIndex(where: { $0.name == field.name }) {
                state.manifest[index] = field
            } else {
                state.manifest.append(field)
            }
            state.lastSentAt.removeValue(forKey: field.name)
            writeState(state, for: serverUrl)
        }
    }

    func removeField(_ name: String, for serverUrl: String) {
        queue.sync {
            guard var state = readState(for: serverUrl), state.enabled else {
                return
            }
            state.manifest.removeAll { $0.name == name }
            state.lastSentAt.removeValue(forKey: name)
            writeState(state, for: serverUrl)
        }
    }

    func updateLastSentAt(_ lastSentAt: [String: Double], for serverUrl: String) {
        queue.sync {
            guard var state = readState(for: serverUrl) else {
                return
            }
            state.lastSentAt = lastSentAt
            writeState(state, for: serverUrl)
        }
    }

    func setStableValues(_ values: [String: String]) {
        queue.sync {
            guard let data = try? JSONEncoder().encode(values) else {
                return
            }
            preferences.set(data, forKey: SessionAttributesConstants.stableValuesKey)
        }
    }

    func getStableValue(_ name: String) -> String? {
        queue.sync {
            guard let data = preferences.object(forKey: SessionAttributesConstants.stableValuesKey) as? Data,
                  let values = try? JSONDecoder().decode([String: String].self, from: data) else {
                return nil
            }
            return values[name]
        }
    }
}
